src/hotspot/share/classfile/classFileParser.cpp

```diff
@@ -6751,6 +6768,20 @@ void ClassFileParser::post_process_parsed_stream(const ClassFileStream* const st
                     _class_name->as_C_string(),
                     InstanceKlass::cast(klass)->external_name()));
       }
+      if (fs.has_restricted_type()) {
+        // descriptor type supposed to be a super type of the restricted type, so after the pre-loading
+        // of the restricted type above, the descriptor type should be loaded at this point
+        Symbol* descriptor_name = _cp->symbol_at(_descriptor_signature_info->at(fs.index()));
+        Klass* desc_klass = SystemDictionary::resolve_or_null(descriptor_name,
+                                                   Handle(THREAD, _loader_data->class_loader()),
+                                                   _protection_domain, CHECK);
+        if (desc_klass == NULL || !klass->is_subtype_of(desc_klass)) {
```

Comment:
Consider if using a ResolvingSignatureStream & a call to ResolvingSignatureStream::as_klass() might be a more JVM consistent way to obtain the Klass from the descriptor signature.